homeassistant/components/influxdb/__init__.py

```diff
@@ -55,98 +65,133 @@
 BATCH_TIMEOUT = 1
 BATCH_BUFFER_SIZE = 100
 
-COMPONENT_CONFIG_SCHEMA_ENTRY = vol.Schema(
-    {vol.Optional(CONF_OVERRIDE_MEASUREMENT): cv.string}
+DB_CONNECTION_FAILURE_MSG = ()
+
+COMPONENT_CONFIG_SCHEMA_CONNECTION = {
+    # Connection config for V1 API.
+    vol.Optional(CONF_HOST): cv.string,
+    vol.Inclusive(CONF_USERNAME, "authentication"): cv.string,
+    vol.Inclusive(CONF_PASSWORD, "authentication"): cv.string,
+    vol.Optional(CONF_DB_NAME, default=DEFAULT_DATABASE): cv.string,
+    vol.Optional(CONF_PATH): cv.string,
+    vol.Optional(CONF_PORT): cv.port,
+    vol.Optional(CONF_SSL): cv.boolean,
+    vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+    # Connection config for V2 API.
+    vol.Inclusive(CONF_API_V2, "v2_api_config"): True,
```

Comment:
Make this an option to specify version in general and not a toggle for v2. If v3 comes out, adding another toggle isn't wished for.

Reply:
I tried that first.  The problem I had was that I couldn't use voluptuous to validate the V2 config that way.  If I had an `api_version` field I couldn't figure out any way to get the config to validate that `url`, `bucket` and `token` were all required when `api_version = 2`.  

Well I should say I did figure out a way but it requires changing from `vol.Schema` to `vol.All` so you can provide a function in addition to a schema.  But it seems that can't be done since `CONFIG_SCHEMA` is going to be extended with the base `PLATFORM_SCHEMA` and has to be a vol.Schema for that to be possible, it can't be a vol.Any.

So the best I could come up with was using `vol.Inclusive` which then demands if `api_v2` has a value then `bucket` and `token` must also have a value. This felt better to me then doing all the config checking in the runtime code and having situations where check config passes but users see errors for `missing bucket/token` on startup.  But yes the downside is that if/when a V3 API came around you'd need to repeat and make an `api_v3` field since it probably has different auth requirements.

I'm open to suggestions.  Is there a better approach here?  I looked around but couldn't find anything similar to use as an example.

Reply:
Not sure how to handle the validation. Added a `second-opinion-wanted` to see if one with more experience on the matter can advise on this.

Reply:
We can use `vol.All` like so:
```py
{
    DOMAIN: vol.All(
        _CONFIG_SCHEMA_PLATFORM.extend(COMPONENT_CONFIG_SCHEMA_CONNECTION),
        version_validator,
    )
}
```
The `version_validator` should check the configured api version and apply the correct schema corresponding to the version to the passed dict.

Reply:
Right sorry, I should've clarified more.  This is exactly what I had and it works great in this spot.  

The problem is the sensor.  In the sensor all these fields are added to `PLATFORM_SCHEMA` and that is then extended as a schema.  When I tried to make `PLATFORM_SCHEMA` into a `vol.All` like this then I ran into errors on startup.  So it seemed like I had to make a choice between two not great options:

1. Have a different way of configuring the connection to the InfluxDB deployment for the sensor then for the main platform
2. Use the `inclusive` approach to validation I took since that works in both places

Reply:
We're not modifying the platform schema here.

Reply:
Use the same pattern as above for the platform schema, ie put `vol.All` around all schemas that represent dicts.

Reply:
Sorry I'm probably still being confusing.  I can absolutely make the changes you're proposing in __init__.py, completely makes sense.  But sensor.py also shares the same config options (`host`, `ssl`, `port`, `bucket`, `token`, etc.) since it also needs to know how to connect to an InfluxDB deployment.

Inside sensor.py, it initializes the `PLATFORM_SCHEMA` object like this:
```python
PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(COMPONENT_CONFIG_SCHEMA_CONNECTION).extend(
    {
        vol.Exclusive(CONF_QUERIES, "queries"): [_QUERY_SCHEMA["InfluxQL"]],
        vol.Exclusive(CONF_QUERIES_FLUX, "queries"): [_QUERY_SCHEMA["Flux"]],
    }
)
```

Notice that `COMPONENT_CONFIG_SCHEMA_CONNECTION` is shared between both __init__.py and sensor.py, that was intended to be the common validator used for validating basic connection details in both places.

So I can make the adjustments you're talking about but only in __init__.py. In sensor.py since the connection details are put into `PLATFORM_SCHEMA` I wouldn't be able to use the   custom validator there. 

What would you recommend doing for the sensor in this case?  Should I continue to rely on `inclusive` there and have the configuration options be different for the sensor vs the recorder?  Or should I make both configurations identical but do additional validation on initialization in the sensor?

Reply:
Alternatively I could adjust the config for the sensor so platform schema looks like this:
```python
PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
    "connection": vol.All(
        COMPONENT_CONFIG_SCHEMA_CONNECTION,
        version_validator,
    ),
    vol.Exclusive(CONF_QUERIES, "queries"): [_QUERY_SCHEMA["InfluxQL"]],
    vol.Exclusive(CONF_QUERIES_FLUX, "queries"): [_QUERY_SCHEMA["Flux"]],
})
```
Since the connection details are now in an object field off of `PLATFORM_SCHEMA` then I would be able to use `vol.All` and `version_validator` but that woudl be a breaking change compared to the v1 config.  Since in v1 all the basic connection fields were top level in the `PLATFORM_SCHEMA` object

Reply:
As I said in the other comment, it seems I misunderstood an earlier error around using `vol.All`.  I've now changed the config to use an `api_version` field and am validating the rest of the config based on whether `api_version` is equal to 1 or 2.  The validation is improved and it provides a pattern to ensure we can easily support new versions of the API in the future 

I need to fix the doc and tests to reflect these new changes still.